File: .agents/skills/ros2-pkg-dev/templates/service_client_example.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
============================================
Python ROS2 Service Client 示例 (rclpy)
============================================
功能: 调用加法计算服务
服务类型: example_interfaces/srv/AddTwoInts
用法: 参考此代码在现有节点中添加 Service Client
============================================
"""

import functools
import random

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


class ServiceClientExample(Node):
    """Service Client 示例节点"""

    def __init__(self):
        super().__init__('service_client_example')

        # 创建 Service Client
        # 参数: 服务类型, 服务名称
        self.client = self.create_client(AddTwoInts, 'add_two_ints')

        # 等待服务上线
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('等待服务 /add_two_ints 上线...')
            if not rclpy.ok():
                self.get_logger().error('被中断，退出')
                return

        self.get_logger().info('Service Client 已启动，服务已连接')

        # 创建定时器，每2秒调用一次服务
        self.timer = self.create_timer(2.0, self.send_request)
        self.call_count = 0

    def send_request(self):
        """发送服务请求"""
        # 创建请求
        request = AddTwoInts.Request()
        request.a = random.randint(10, 99)
        request.b = random.randint(10, 99)

        self.get_logger().info(f'发送请求: a={request.a}, b={request.b}')

        # 异步发送请求
        future = self.client.call_async(request)

        # 使用 functools.partial 绑定额外参数到回调函数
        # 比 lambda 更清晰，且能正确处理异常
        future.add_done_callback(
            functools.partial(self._handle_response, request.a, request.b))

        self.call_count += 1
        # 示例：只调用5次
        if self.call_count >= 5:
            self.timer.cancel()
            self.get_logger().info('已完成5次调用，停止定时器')

    def _handle_response(self, a: int, b: int, future):
        """处理服务响应
        
        Args:
            a: 请求的第一个数（通过 partial 绑定）
            b: 请求的第二个数（通过 partial 绑定）
            future: 包含响应结果的 Future 对象（由 ROS2 传入）
        """
        try:
            response = future.result()
            self.get_logger().info(f'收到响应: {a} + {b} = {response.sum}')
        except Exception as e:
            self.get_logger().error(f'服务调用失败: {str(e)}')

    def destroy_node(self):
        """销毁节点"""
        self.get_logger().info('正在销毁 Service Client 节点...')
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = ServiceClientExample()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('收到键盘中断')
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: .agents/skills/ros2-pkg-dev/templates/service_client_example.cpp
// ============================================
// C++ ROS2 Service Client 示例 (rclcpp)
// ============================================
// 功能: 调用加法计算服务
// 服务类型: example_interfaces/srv/AddTwoInts
// 用法: 参考此代码在现有节点中添加 Service Client
// ============================================

#include <rclcpp/rclcpp.hpp>
#include <example_interfaces/srv/add_two_ints.hpp>

#include <chrono>

using namespace std::chrono_literals;

class ServiceClientExample : public rclcpp::Node
{
public:
  ServiceClientExample()
  : Node("service_client_example")
  {
    // 创建 Service Client
    // 参数: 服务名称
    client_ = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");

    // 等待服务上线
    while (!client_->wait_for_service(1s)) {
      if (!rclcpp::ok()) {
        RCLCPP_ERROR(this->get_logger(), "被中断，退出");
        return;
      }
      RCLCPP_INFO(this->get_logger(), "等待服务 /add_two_ints 上线...");
    }

    RCLCPP_INFO(this->get_logger(), "Service Client 已启动，等待调用服务");

    // 示例：定时调用服务
    timer_ = this->create_wall_timer(
      2s, std::bind(&ServiceClientExample::send_request, this));
  }

private:
  void send_request()
  {
    // 判断服务是否就绪
    if (!client_->service_is_ready()) {
      RCLCPP_WARN(this->get_logger(), "服务 /add_two_ints 当前不可用，跳过本次请求");
      return;
    }

    // 创建请求
    auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    request->a = 10 + rand() % 90;  // 随机数 10-99
    request->b = 10 + rand() % 90;

    RCLCPP_INFO(this->get_logger(), "发送请求: a=%ld, b=%ld", request->a, request->b);

    // 异步发送请求
    auto result_future = client_->async_send_request(request);

    // 等待响应（阻塞方式，可选）
    // auto result = rclcpp::spin_until_future_complete(shared_from_this(), result_future);

    // 非阻塞方式：使用回调
    result_future.then(
      [this, request](rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future) {
        try {
          auto response = future.get();
          RCLCPP_INFO(this->get_logger(), 
            "收到响应: %ld + %ld = %ld", request->a, request->b, response->sum);
        } catch (const std::exception &e) {
          RCLCPP_ERROR(this->get_logger(), "服务调用失败: %s", e.what());
        }
      });
  }

  // Service Client 成员变量
  rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ServiceClientExample>());
  rclcpp::shutdown();
  return 0;
}
